paymentsheet/src/main/java/com/stripe/android/paymentsheet/BaseAddPaymentMethodFragment.kt

```diff
@@ -5,249 +5,302 @@ import android.view.LayoutInflater
-            )
-            replace(
-                R.id.payment_method_fragment_container,
-                fragmentForPaymentMethod(paymentMethod),
-                args
-            )
-        }
+    private fun onLinkSignupStateChanged(
+        config: LinkPaymentLauncher.Configuration,
+        viewState: InlineSignupViewState
+    ) {
+        sheetViewModel.updatePrimaryButtonUIState(
+            if (viewState.useLink) {
+                val userInput = viewState.userInput
+                if (userInput != null &&
+                    sheetViewModel.selection.value != null
+                ) {
+                    PrimaryButton.UIState(
+                        label = null,
+                        onClick = {
+                            sheetViewModel.payWithLinkInline(
+                                config,
+                                userInput
+                            )
+                        },
+                        enabled = true,
+                        visible = true
+                    )
+                } else {
+                    PrimaryButton.UIState(
+                        label = null,
+                        onClick = null,
+                        enabled = false,
+                        visible = true
+                    )
+                }
+            } else {
+                null
+            }
+        )
     }
 
-    private fun updateLinkInlineSignupVisibility(selectedPaymentMethod: SupportedPaymentMethod) {
-        lifecycleScope.launch {
-            showLinkInlineSignup.value = sheetViewModel.isLinkEnabled.value == true &&
-                sheetViewModel.stripeIntent.value
-                    ?.linkFundingSources?.contains(PaymentMethod.Type.Card.code) ?: false &&
-                selectedPaymentMethod.code == PaymentMethod.Type.Card.code &&
-                sheetViewModel.linkConfiguration.value
-                ?.let {
-                    sheetViewModel.linkLauncher.getAccountStatusFlow(it).first()
-                } == AccountStatus.SignedOut
+    @VisibleForTesting
+    internal fun createFormArguments(
+        selectedItem: LpmRepository.SupportedPaymentMethod,
+        showLinkInlineSignup: Boolean
+    ) = getFormArguments(
+        showPaymentMethod = selectedItem,
+        stripeIntent = requireNotNull(sheetViewModel.stripeIntent.value),
+        config = sheetViewModel.config,
+        merchantName = sheetViewModel.merchantName,
+        amount = sheetViewModel.amount.value,
+        injectorKey = sheetViewModel.injectorKey,
+        newLpm = sheetViewModel.newPaymentSelection,
+        isShowingLinkInlineSignup = showLinkInlineSignup
+    )
 
-            viewBinding.linkInlineSignup.isVisible = showLinkInlineSignup.value
-        }
-    }
+    @VisibleForTesting
+    internal fun transformToPaymentSelection(
+        formFieldValues: FormFieldValues?,
+        selectedPaymentMethodResources: LpmRepository.SupportedPaymentMethod
+    ) = formFieldValues?.let {
```

Comment:
I find that this should eventually live outside the fragment, so another follow-up improvement opportunity.